File: src/android/security/src/main/java/org/meshchat/identity/StaffKeyVault.kt
package org.meshchat.identity

import android.content.Context

/** One staff credential encrypted by its own OS wrapping key, bound to this
 * device identity generation. No unwrapped key/session survives an operation. */
class StaffKeyVault internal constructor(private val files: IdentityStorage, private val protection: IdentityProtection) : IdentityResetStore {
    companion object {
        fun android(context: Context): StaffKeyVault = StaffKeyVault(
            AndroidIdentityStorage(context.applicationContext,"meshchat-staff-v1"),
            AndroidIdentityProtection(context.applicationContext,"org.meshchat.staff.wrapping.v1",1..512))
    }
    private fun refuse(failure: IdentityFailure): Nothing = throw IdentityProviderException(failure)
    fun present(): Boolean = IdentityProvider.withOperation {
        files.exists(IdentityFile.ENVELOPE)||files.exists(IdentityFile.JOURNAL)||protection.exists()
    }
    private fun header(generation: ByteArray): ByteArray {
        if(generation.size!=16)refuse(IdentityFailure.INVALID_INPUT)
        return byteArrayOf(1)+generation
    }
    /** Validation runs before any mutation. Atomic replacement preserves the old
     * encrypted credential if validation, encryption or the file write fails. */
    fun importConfirmed(generation: ByteArray, candidate: ByteArray, validate: (String)->Unit) = IdentityProvider.withOperation {
        try {
            protection.requireUnlocked()
            if(candidate.size !in 1..512)refuse(IdentityFailure.INVALID_INPUT)
            val aad=header(generation)
            if(files.exists(IdentityFile.JOURNAL))refuse(IdentityFailure.RECOVERY_REQUIRED)
            validate(candidate.toString(Charsets.UTF_8))
            val existing=files.exists(IdentityFile.ENVELOPE)
            if(existing && !protection.exists())refuse(IdentityFailure.INVALIDATED)
            if(!existing) {
                if(protection.exists())refuse(IdentityFailure.RECOVERY_REQUIRED)
                files.write(IdentityFile.JOURNAL,byteArrayOf(1));protection.create()
            }
            val encrypted=protection.seal(candidate,aad)
            files.write(IdentityFile.ENVELOPE,aad+encrypted)
            if(!existing)files.delete(IdentityFile.JOURNAL)
        } finally { candidate.fill(0) }
    }
    fun <T> access(generation: ByteArray, work: (String)->T): T = IdentityProvider.withOperation {
        protection.requireUnlocked()
        if(files.exists(IdentityFile.JOURNAL))refuse(IdentityFailure.RECOVERY_REQUIRED)
        if(!files.exists(IdentityFile.ENVELOPE))refuse(if(protection.exists())IdentityFailure.RECOVERY_REQUIRED else IdentityFailure.MISSING)
        if(!protection.exists())refuse(IdentityFailure.INVALIDATED)
        val aad=header(generation);val encoded=files.read(IdentityFile.ENVELOPE)
        if(encoded.size !in 46..557 || !encoded.copyOfRange(0,17).contentEquals(aad))refuse(IdentityFailure.INVALID_INPUT)
        val plain=protection.open(encoded.copyOfRange(17,encoded.size),aad)
        try { if(plain.size !in 1..512)refuse(IdentityFailure.INVALID_INPUT);val result=work(plain.toString(Charsets.UTF_8));protection.requireUnlocked();result }
        finally {plain.fill(0)}
    }
    fun forget() = IdentityProvider.withOperation {
        protection.requireUnlocked()
        files.write(IdentityFile.JOURNAL,byteArrayOf(2))
        protection.delete();files.delete(IdentityFile.ENVELOPE);files.delete(IdentityFile.JOURNAL)
    }
    override fun clearIdentityState() { IdentityProvider.requireResetInProgress();forget() }
}

File: src/ios/Security/StaffKeyVault.swift
import Foundation

/// Main-actor serialization includes the final native submission. No await or
/// retained private session can cross the foreground/lock/reset boundary.
@MainActor final class StaffKeyVault: IdentityResetStore {
    private let files: IdentityStorage
    private let protection: IdentityProtection
    private(set) var foreground = true
    init(files: IdentityStorage, protection: IdentityProtection) {
        self.files = files; self.protection = protection
    }
    static func native() throws -> StaffKeyVault {
        StaffKeyVault(files: try AppleIdentityStorage(name: "meshchat-staff-v1"),
                      protection: AppleIdentityProtection(tag: "org.meshchat.staff.wrapping.v1", payloadRange: 1...512))
    }
    func setForeground(_ value: Bool) { foreground = value }
    func present() throws -> Bool { try files.exists(.envelope) || files.exists(.journal) || protection.exists() }
    private func header(_ generation: Data) throws -> Data {
        guard generation.count == 16 else { throw IdentityFailure.invalidInput }
        return Data([1]) + generation
    }
    private func requireForeground() throws {
        try protection.requireUnlocked()
        guard foreground else { throw OrganizerError.Authority }
    }
    func importConfirmed(generation: Data, candidate: inout Data, validate: (String) throws -> Void) throws {
        defer { candidate.resetBytes(in: 0..<candidate.count); candidate.removeAll() }
        try requireForeground()
        guard (1...512).contains(candidate.count), let uri = String(data: candidate, encoding: .utf8) else { throw IdentityFailure.invalidInput }
        let aad = try header(generation)
        guard !files.exists(.journal) else { throw IdentityFailure.recoveryRequired }
        try validate(uri)
        let existing = files.exists(.envelope)
        if existing { guard try protection.exists() else { throw IdentityFailure.invalidated } }
        else {
            guard try !protection.exists() else { throw IdentityFailure.recoveryRequired }
            try files.write(.journal, Data([1])); try protection.create()
        }
        try files.write(.envelope, aad + protection.seal(candidate, aad: aad))
        if !existing { try files.delete(.journal) }
    }
    func access<T>(generation: Data, _ work: (String) throws -> T) throws -> T {
        try requireForeground()
        guard !files.exists(.journal) else { throw IdentityFailure.recoveryRequired }
        guard files.exists(.envelope) else { throw IdentityFailure.missing }
        guard try protection.exists() else { throw IdentityFailure.invalidated }
        let aad = try header(generation), encoded = try files.read(.envelope)
        guard (18...1024).contains(encoded.count), encoded.prefix(17) == aad else { throw IdentityFailure.invalidInput }
        var plain = try protection.open(Data(encoded.dropFirst(17)), aad: aad)
        defer { plain.resetBytes(in: 0..<plain.count) }
        guard (1...512).contains(plain.count), let uri = String(data: plain, encoding: .utf8) else { throw IdentityFailure.invalidInput }
        let result = try work(uri); try requireForeground(); return result
    }
    func forget() throws {
        try protection.requireUnlocked()
        try files.write(.journal, Data([2])); try protection.delete()
        try files.delete(.envelope); try files.delete(.journal)
    }
    func clearIdentityState() throws { try IdentityProvider.requireResetInProgress(); try forget() }
}

@MainActor final class FeatureResetStore: IdentityResetStore {
    private let storage: StorageVault
    private let staff: StaffKeyVault
    init(storage: StorageVault, staff: StaffKeyVault) { self.storage = storage; self.staff = staff }
    func clearIdentityState() throws { try staff.clearIdentityState(); try storage.clearIdentityState() }
}
